---
title: "Projects in Spatial Transcriptomics"
output: 
  html_document:
    keep_md: true
---

Load count data for breast cancer layer 1. For each of 254 locations on the tissue, load counts of 14,881 genes expressed at that location.  Source data from "Visualization and analysis of gene expression in tissue sections by spatial transcriptomics" (Stahl, Patrik, et al 2016). 

```{r}
st <- read.csv('Layer1_BC_count_matrix-1.tsv',header=TRUE,sep="\t")
dim(st)
```

Counts for 14,881 genes are listed for each of 254 locations:
```{r}
head(st[1:5,1:6])
```

The first column lists the coordinates of the locations. 
```{r}
coordinates <- st$X
x <- as.numeric(substr(coordinates,1,as.integer(regexpr("x",coordinates))-1))
y <- as.numeric(substr(coordinates,as.integer(regexpr("x",coordinates))+1,nchar(as.character(coordinates))))

#head(coordinates)
#head(x)
#head(y)
#plot(x,y)
plot(x * 288.9, y * 292.6,ylim=c(0,9000),xlim=c(0,9000)) #upscale coordinates (or can resize image down to 33x35)
```


# Using T-SNE to cluster locations
```{r}
library(Rtsne)
set.seed(1989)
#st_mat <- as.matrix(st[,-1])
test2 <- Rtsne(st,dim=3)
#sample_names <- st[,1]
#test <- normalize_input(st_mat)
#dim(test)
#head(test[1:5,1:5])
#dim(st)
#head(st)
```

Cluster first 2 dimension results from T-SNE, using k-means(7)
```{r}
set.seed(1989)
clusters <- kmeans(test2$Y,7)$cluster
plot(test2$Y,col=clusters,asp=1)
```

# Overlay original tissue slide
Load stained image of breast cancer layer 1 (source: https://www.spatialresearch.org/resources-published-datasets/)
```{r}
library(magick)
full_img <- image_read('README_files/screenshots/HE_layer1_BC.jpg')
image_info(full_img)
```

Downscale and preview image:
```{r}
plot(image_scale(full_img, "1000"))
```


```{r}
overlay <- image_scale(full_img, "1000")
overlay_small <- image_scale(full_img, "1000")
cols <- c("red","green","blue","yellow","lime","magenta","wheat4")
for (i in seq(1,length(x))) {
  xi <- x[i] * 288.9 * (1000/9245) #upscale to 9k x 9k, then downscale to 1k x 1k
  yi <- y[i] * 292.6 * (1076/9949)
  locationi <- paste("+",as.character(xi),"+",as.character(yi),sep="")
  overlay <- image_annotate(overlay, "•", size = 150, gravity = "southwest", color = cols[clusters[i]], location=locationi)
  overlay_small <- image_annotate(overlay_small, "•", size = 75, gravity = "southwest", color = cols[clusters[i]], location=locationi)
}
plot(overlay_small)
```

The clusters can be used in the plot of x-y location coordinates from the tissue sample. (Save as a gif)
```{r}
test <- image_scale(c(full_img,overlay_small), "1000")
image_write(image_animate(image_morph(test, 10)), "README_files/screenshots/bc1.gif")
```

#![tissue slide](README_files/screenshots/bc1.gif)


# Resources
* Source: Ståhl, Patrik L., et al. "Visualization and analysis of gene expression in tissue sections by spatial transcriptomics." Science 353.6294 (2016): 78-82.
  * Source data: https://www.spatialresearch.org/resources-published-datasets/

* Trendsceek, by Daniel Edsgärd. Vignette: 
  * https://github.com/edsgard/trendsceek/blob/master/vignettes/trendsceek.Rmd
  * Source Paper? "Identification of spatial expression trends in single-cell gene expression data"
    * https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6314435/
    
* Gene expression atlas of a developing tissue by single cell expression correlation analysis
  * https://www.nature.com/articles/s41592-019-0492-x
  * Data here: https://www.ncbi.nlm.nih.gov/search/all/?term=GSE127832

* Subsample large sc datasets: https://www.cell.com/cell-systems/pdfExtended/S2405-4712(19)30152-8





# Future
Computational geometry. 
https://www.cgal.org/projects.html#astronomy

One featured example that uses that package: taking redshift data and creating structure. Could perform similar operation to expression data? To create compartments of tissue based on different cell types in each subtissue component...
#![example "adhesive" algorithm](README_files/screenshots/example.computational.geometry.png)



