---
title: "May 25th Test, Compare scRNA with one spatial sample"
output: md_document
---

# Compare scRNA with one spatial sample
May 25th, 2020. Dive into the data for a specific single cell and a single spatial sample which was enriched for that cell's cell type.

```{r}
library(Seurat)
library(tidyr)
library(ggplot2)
library(igraph)
```

## Identify a target spatial sample
The spatial data is available publically from 10x Genomics. 1,072 spatially tagged samples for 47,094 genes.
```{r}
cortex_final <- readRDS('../2/cortex_final.rds') #342MB
```
```{r}
cortex_final
#names(cortex_final) 
#spatial is the base data, sct is the normalized data, predictions are the anchor-based integration results, pca & umap dim reduction results, anterior1 are the tissue image and mapping tables
```

There are 70 spatial samples that were placed in cluster 1 in an earlier analysis. The Oligo celltype seems to be the dominant label for most, but not all, of the cluster 1 samples. Median anchor-based integration output for cluster 1 cells is .9933 (out of 1.0).
```{r}
cl <- 1
target.celltype <- "Oligo"
x <- as.data.frame(cortex_final[['predictions']]@data[-24,cortex_final$seurat_clusters==cl])
#dim(x)
summary(t(x[rownames(x)==target.celltype,]))
#boxplot(x[rownames(x)==target.celltype,])
```

Our randomly selected cluster 1 sample is "TGGCAGCAGTAATAGT-1", which was predicted to be componsed of .9949 Oligo celltype.
```{r}
set.seed(1989)
target.sample <- sample(colnames(x),1)
target.sample #"TGGCAGCAGTAATAGT-1"
x[,colnames(x)==target.sample]
```

That spatial sample contains 16,017 normalized gene expression levels between 0.0 and 7.35:
```{r}
m <- which(cortex_final[['SCT']]@data@Dimnames[[2]] == target.sample) #our target sample is the 974th out of 1,072 in the SCT & Spatial assays
spatial.sample <- as.data.frame(cortex_final[['SCT']]@data[,m]) #to save space, could always set row.names = NULL and get gene names from the original seurat object
dim(spatial.sample) #16,017 genes for this one transformed spot
colnames(spatial.sample) <- 'SCT'
#spatial.sample <- cbind(spatial.sample, as.data.frame(cortex_final[['Spatial']]@data[,m])) #would like to see how they normalized from Spatial assay to SCT assay. Cant do a simble cbind because the original Spatial data had 31053 genes but only 16017 are included in the cleaned SCT assay
summary(spatial.sample)
```

## Identify highly expressed genes
Out of 16,017 genes, 11,092 were zero [69%] and 4,925 were above zero [31%].

```{r}
sum(spatial.sample == 0)
sum(spatial.sample > 0)
spatial.sample.nonzeros <- as.data.frame(spatial.sample[spatial.sample > 0,])
colnames(spatial.sample.nonzeros) <- 'SCT'
summary(spatial.sample.nonzeros)

```

Majority of the SCT (normalized) expressions for this sample were below 3.
```{r}
ggplot(spatial.sample.nonzeros, aes("",SCT)) + 
  geom_violin() + #couldnt get seurat VlnPlot to cooperate
  scale_y_continuous('transformed expression level', seq(0,7,1)) +
  scale_x_discrete(paste("nonzero expressions for spatial sample:",target.sample))
```

Lets bin the data for the expressions from the spatial sample and see how many genes are in each bin:
```{r}
breaks <- c(seq(0,3,.25), seq(4,8,1))
spatial.sample.bins <- cut(as.numeric(spatial.sample[,1]), breaks, include.lowest = TRUE)
table(spatial.sample.bins)
```

There are only a 34 genes with expression over 4.
```{r}
target.high.expr.genes <- row.names(spatial.sample)[spatial.sample > 4]
length(target.high.expr.genes)
```

Is 34 genes over expression 4 a lot? Lets compare to the other spatial samples... The median number of genes over normalized expression level 4 is 27 [IQR 25,29]
```{r}
x <- colSums(as.matrix(cortex_final[['SCT']]@data>4))
length(x)
summary(x)
```

Is 34 genes over expression 4 a lot for this particular cluster [1]? Not necessarily, as clusters 1 & 2 have a higher mean number of 4.0+ genes. *Not sure if the cell density of sample spots affects the gene counts, or if that has been claculated out in the normalization process.*
```{r}
#identical(names(x), names(cortex_final$seurat_clusters)) #samples are listed in the same order
x <- cbind.data.frame(x,cortex_final$seurat_clusters)
dim(x)
names(x) <- c('Number of genes over 4.0','seurat cluster')
ggplot(x, aes(x=`seurat cluster`,y=`Number of genes over 4.0`)) + 
  geom_jitter(aes(color='blue'),alpha=0.2) +
  geom_boxplot(fill="bisque",color="black",alpha=0.3) + 
  #labs(x='mean education per house') +
  guides(color=FALSE) +
  theme_minimal() 
```

So are there genes that are just high in all clusters? Or are some genes only high in some clusters (can do chi squared test for enrichment?)

Looking at only the first 4 high-epxression genes from the target sample, some are higher across the whole tissue, while others are more localized in high expression:
```{r}
target.high.expr.genes
DefaultAssay(cortex_final) <- "SCT"
SpatialPlot(cortex_final, images='anterior1', features=target.high.expr.genes[1:4], ncol=2)
```


## For each of the 34 high expr genes... take one gene, compare its level to its level in the neighboring spots?
Can turn this one-gene-level data into a network graph, segmentation on it? Look for areas with a steep drop off?? Could do the same with celltype predictions too.
Plp1 or Tmsb4x look like a good candidates because they have a wide range of levels in the slide. 
```{r}
d <- data.frame(cortex_final[['anterior1']]@coordinates[,c('col','row')])
d['row'] <- d['row'] * -1
colnames(d) <- c('x','y')
d <- cbind(d,t(as.matrix((cortex_final[['predictions']]@data))))
dim(d)
d[1:10,1:5]
```

## Convert spatial samples into a network graph
Manually set a cutoff Euclidean distance of 2.1
```{r}
test <- as.matrix(dist(data.frame(d[,1:2])))
typeof(test)
dim(test)

test[1:10,1:5]

hist(colSums((test)<2.1))

#count(which(as.numeric(test[,1]) <= 20))
#hist(colSums(test>0 & test<6))

test <- test>0 & test<2.1 #close neighbors only
hist(colSums(test))

```

```{r, fig.width = 12, fig.height = 12}
g <- graph_from_adjacency_matrix(test, mode="undirected")
 #iGraph plotting help: https://kateto.net/netscix2016.html
l <- as.matrix(d[,1:2])
plot(g, vertex.label=NA, vertex.size=1, layout = l) #layout_with_kk is pretty good too
```





### Verify that our target sample's neighbors are properly accounted for in the conversion to an adjaceny matrix and then to a network graph...
our target sample is at 61,-39
```{r}
d[target.sample,c('x','y')] #our target sample is at 61,-39
```

Neighbors:
```{r}
d[58 < d$x & d$x < 64 & -42 < d$y & d$y < -36,c('x','y')]

```

```{r}
d$color <- as.numeric(58 < d$x & d$x < 64 & -42 < d$y & d$y < -36) #the neighbors
d[target.sample,'color'] <- 9 #our target
table(d$color)
ggplot(d,aes(x,y,color = as.factor(color)))+
  geom_point()
```

Verify the 8 manually identified neighbors were all included in the automatically identified neighbors from the distance matrix:
```{r}
which(test[,target.sample]) #the adjacency matrix identified 5 neighbors that met the maximum distance threshold
names(which(test[,target.sample])) %in% rownames(d[58 < d$x & d$x < 64 & -42 < d$y & d$y < -36,c('x','y')]) #all 5 were in the group of 8 manually identified neighbors
```



```{r}
rm(cortex_final) #remove spatial cortex object from memory
```




